### PR TITLE
New feature: cross-team search

### PR DESCRIPTION
**Pull request by smatting**
This PR allows teams to be searchable (fuzzy search on handle and name) by users from others teams. By default the behaviour is a pre-PR: teams are not generally searchable, except if the searcher provides the exact handle as search query.

https://wearezeta.atlassian.net/browse/FS-491

 - [x] Test  `elasticsearch-index` chart: test automated data migration procedure

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] If HTTP endpoint paths have been added or renamed, the **endpoint / config-flag checklist** (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.
 - [x] If a cassandra schema migration has been added, I ran **`make git-add-cassandra-schema`** to update the cassandra schema documentation.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.
   - [x] If new config options introduced: added usage description under docs/reference/config-options.md
   - [x] If new config options introduced: recommended measures to be taken by on-premise instance operators.

### CODE REVIEW

File: docs/reference/cassandra-schema.cql [hunk 1]
```diff
@@ -492,6 +493,101 @@ CREATE TABLE galley_test.team_member (
...
+    AND crc_check_chance = 1.0
+    AND dclocal_read_repair_chance = 0.1
+    AND default_time_to_live = 0
+    AND gc_grace_seconds = 864000
+    AND max_index_interval = 2048
+    AND memtable_flush_period_in_ms = 0
+    AND min_index_interval = 128
+    AND read_repair_chance = 0.0
+    AND speculative_retry = '99PERCENTILE';
+
```
Comment (akshaymankar) [inline docs/reference/cassandra-schema.cql:590]: Looks like your local cassandra has more keyspaces, I think you should delete them and regenerate the file.

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs [hunk 2]
```diff
@@ -1033,6 +1036,15 @@ type AllFeatureConfigsGet =
         :> Get '[Servant.JSON] AllFeatureConfigs
     )
 
+type FeatureNoConfigMultiGetBase featureName =
+  Summary
+    (AppendSymbol "Get team feature status in bulk for feature " (KnownTeamFeatureNameSymbol featureName))
+    :> Capture "tid" TeamId
```
Comment (akshaymankar) [inline libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:1042]:
```suggestion
```

Shouldn't be required, right?

Comment (smatting) [reply libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:1042]: Yes, it's copypasta leftover

File: services/brig/src/Brig/User/Search/SearchIndex.hs [hunk 3]
```diff
@@ -184,30 +184,60 @@ termQ f v =
...
+restrictSearchSpace :: SearchSetting -> ES.Query
+restrictSearchSpace FederatedSearch =
+  ES.QueryBoolQuery
+    boolQuery
+      { ES.boolQueryShouldMatch =
+          [ matchNonTeamMemberUsers,
+            matchTeamMembersSearchableByAllTeams
+          ]
+      }
+restrictSearchSpace (LocalSearch _uid mteam searchInfo) =
```
Comment (battermann) [inline services/brig/src/Brig/User/Search/SearchIndex.hs:201]:
```suggestion
restrictSearchSpace (LocalSearch _ mteam searchInfo) =
```

File: services/brig/test/integration/API/Search.hs [hunk 4]
```diff
@@ -87,27 +89,38 @@ tests opts mgr galley brig = do
...
+                testGroup "order" $
+                  [ test mgr "team-mates are listed before team-outsiders (exact match)" $ testSearchOrderingAsTeamMemberExactMatch brig,
+                    test mgr "team-mates are listed before team-outsiders (prefix match)" $ testSearchOrderingAsTeamMemberPrefixMatch brig,
+                    test mgr "team-mates are listed before team-outsiders (worse name match)" $ testSearchOrderingAsTeamMemberWorseNameMatch brig,
+                    test mgr "team-mates are listed after team-outsiders (worse handle match)" $ testSearchOrderingAsTeamMemberWorseHandleMatch brig
+                  ]
+              ],
+            testGroup "when SearchVisibilityInbound == SearchableByAllTeams" $
+              [ test mgr "team member cannot be found by non-team user with display name" $ testSearchTeamMemberAsNonMemberDisplayName mgr brig galley TeamFeatureEnabled,
+                test mgr "team member can be found by non-team user with exact handle" $ testSearchTeamMemeberAsNonMemberExactHandle mgr brig galley TeamFeatureEnabled,
```
Comment (battermann) [inline services/brig/test/integration/API/Search.hs:108]:
```suggestion
                test mgr "team member can be found by non-team user with exact handle" $ testSearchTeamMemberAsNonMemberExactHandle mgr brig galley TeamFeatureEnabled,
```

File: services/brig/test/integration/API/Search.hs [hunk 5]
```diff
@@ -87,27 +89,38 @@ tests opts mgr galley brig = do
...
-            test mgr "team A member cannot be found by team B member" $ testSearchTeamMemberAsOtherMemberOutboundOnly brig testSetupOutboundOnly,
-            test mgr "team A member *can* be found by other team A member" $ testSearchTeamMemberAsSameMemberOutboundOnly brig testSetupOutboundOnly,
-            test mgr "non team user cannot be found by a team member A" $ testSeachNonMemberAsTeamMemberOutboundOnly brig testSetupOutboundOnly
+          [ testGroup "When SearchVisibilityInbound == SearchableByOwnTeam" $
+              [ test mgr "team member cannot be found by non-team user" $ testSearchTeamMemberAsNonMemberOutboundOnly brig testSetupOutboundOnly,
+                test mgr "team A member cannot be found by team B member" $ testSearchTeamMemberAsOtherMemberOutboundOnly brig testSetupOutboundOnly,
+                test mgr "team A member *can* be found by other team A member" $ testSearchTeamMemberAsSameMemberOutboundOnly brig testSetupOutboundOnly,
+                test mgr "non team user cannot be found by a team member A" $ testSeachNonMemberAsTeamMemberOutboundOnly brig testSetupOutboundOnly
```
Comment (battermann) [inline services/brig/test/integration/API/Search.hs:122]:
```suggestion
                test mgr "non team user cannot be found by a team member A" $ testSearchNonMemberAsTeamMemberOutboundOnly brig testSetupOutboundOnly
```

File: services/galley/src/Galley/API/Internal.hs [hunk 6]
```diff
@@ -188,6 +191,19 @@ type IFeatureStatusWithLock f =
...
 
+type FeatureNoConfigMultiGetBase featureName =
+  Summary
+    (AppendSymbol "Get team feature status in bulk for feature " (KnownTeamFeatureNameSymbol featureName))
+    :> "features-multi-teams"
+    :> KnownTeamFeatureNameSymbol featureName
+    :> ReqBody '[Servant.JSON] TeamFeatureNoConfigMultiRequest
+    :> Post '[Servant.JSON] (TeamFeatureNoConfigMultiResponse featureName)
+
+type IFeatureNoConfigMultiGet f =
```
Comment (battermann) [inline services/galley/src/Galley/API/Internal.hs:202]: I would prefer a non-generic type here, if there is no proven need for the generic version.

File: services/galley/src/Galley/API/Teams/Features.hs [hunk 7]
```diff
@@ -757,6 +764,70 @@ setSndFactorPasswordChallengeInternal tid status = do
...
+  TeamId ->
+  Public.TeamFeatureStatus 'Public.WithoutLockStatus 'Public.TeamFeatureSearchVisibilityInbound ->
+  Sem r (Public.TeamFeatureStatus 'Public.WithoutLockStatus 'Public.TeamFeatureSearchVisibilityInbound)
+setTeamSearchVisibilityInboundInternal tid status = do
+  updatedStatus <- setFeatureStatusNoConfig @'Public.TeamFeatureSearchVisibilityInbound (\_ _ -> pure ()) tid status
+  mPersistedStatus <- listToMaybe <$> TeamFeatures.getFeatureStatusNoConfigMulti (Proxy @'Public.TeamFeatureSearchVisibilityInbound) [tid]
+  case mPersistedStatus of
+    Just (pesistedTid, persistedStatus, persistedWriteTime) ->
+      updateSearchVisibilityInbound $
+        Multi.TeamStatusUpdate pesistedTid persistedStatus persistedWriteTime
```
Comment (battermann) [inline services/galley/src/Galley/API/Teams/Features.hs:787]:
```suggestion
    Just (persistedTid, persistedStatus, persistedWriteTime) ->
      updateSearchVisibilityInbound $
        Multi.TeamStatusUpdate persistedTid persistedStatus persistedWriteTime
```

File: services/galley/src/Galley/API/Teams/Features.hs [hunk 8]
```diff
@@ -757,6 +764,70 @@ setSndFactorPasswordChallengeInternal tid status = do
...
+    -- TODO: throw nicer error
```
Comment (battermann) [inline services/galley/src/Galley/API/Teams/Features.hs:788]: fix todo

**Review by battermann — COMMENTED** [review]
LGTM overall, only some minor remarks
don't forget to update `cassandra-schema.cql`

**Review by battermann — APPROVED** [review]

**Pull request merged**